core/src/main/java/org/apache/calcite/rel/metadata/RelMdExpressionLineage.java

```diff
@@ -258,11 +262,17 @@ protected RelMdExpressionLineage() {}
         }
         // Right input references might need to be updated if there are
         // table names clashes with left input
+        final RelDataType fullRowType = SqlValidatorUtil.createJoinType(
+            rexBuilder.getTypeFactory(),
+            rel.getLeft().getRowType(),
+            rel.getRight().getRowType(),
+            null,
+            ImmutableList.of());
```

Comment:
Can you explain this?

Reply:
Because `SEMI`(and `ANTI`) row type does not contains fields from right input, use a full row type here will allow us to inferrer lineage for expressions that reference the right fields.